### PR TITLE
Incremental compilation progress

### PR DESCRIPTION
**Pull request by mlugg**
> Say you build quick? Baby I know,
> That's that incremental

-Sabrina Carpenter 2024

---

We're getting close now, folks!

This PR makes progress towards incremental compilation. Various bugs and regressions have been fixed. The `incr-check` tool is expanded with some more options, such as passing through `-fno-emit-bin` to the compiler. With this option, the existing incremental compilation test, and the 3 added by this commit, all succeed. I don't know whether the right things are going to codegen, but some hacky CBE testing indicates... maybe?

The biggest blocker right now is linker work from @kubkon -- it's currently hard for me to know for sure whether the full pipeline is actually working correctly. Once one of the self-hosted linkers (presumably ELF) can perform basic incremental updates, we can ramp up the testing.

Trying to run the `hello` test without `-fno-emit-bin` currently results in the following linker crash:
```
error: update 'change the string' failed:
thread 16224 panic: reached unreachable code
/home/mlugg/zig/the-great-decl-split-mk2/lib/std/debug.zig:399:14: 0x18c9b9d in assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
/home/mlugg/zig/the-great-decl-split-mk2/lib/std/hash_map.zig:1089:19: 0x1ce013e in putAssumeCapacityNoClobberContext (zig)
            assert(!self.containsContext(key, ctx));
                  ^
/home/mlugg/zig/the-great-decl-split-mk2/lib/std/hash_map.zig:1065:51: 0x1aeeb02 in putNoClobberContext (zig)
            self.putAssumeCapacityNoClobberContext(key, value, ctx);
                                                  ^
/home/mlugg/zig/the-great-decl-split-mk2/lib/std/hash_map.zig:1057:44: 0x193b813 in putNoClobber (zig)
            return self.putNoClobberContext(allocator, key, value, undefined);
                                           ^
/home/mlugg/zig/the-great-decl-split-mk2/src/link/Elf.zig:4175:53: 0x210e222 in allocateAllocSections (zig)
            try self.phdr_to_shdr_table.putNoClobber(gpa, shndx, phndx);
                                                    ^
/home/mlugg/zig/the-great-decl-split-mk2/src/link/Elf.zig:1309:35: 0x1e1b785 in flushModule (zig)
    try self.allocateAllocSections();
                                  ^
/home/mlugg/zig/the-great-decl-split-mk2/src/link/Elf.zig:1008:25: 0x1ba8d8d in flush (zig)
    try self.flushModule(arena, tid, prog_node);
                        ^
/home/mlugg/zig/the-great-decl-split-mk2/src/link.zig:610:77: 0x19c4382 in flush (zig)
                return @as(*tag.Type(), @fieldParentPtr("base", base)).flush(arena, tid, prog_node);
                                                                            ^
/home/mlugg/zig/the-great-decl-split-mk2/src/Compilation.zig:2444:17: 0x19c3bea in flush (zig)
        lf.flush(arena, tid, prog_node) catch |err| switch (err) {
                ^
/home/mlugg/zig/the-great-decl-split-mk2/src/Compilation.zig:2411:22: 0x19c7d96 in update (zig)
            try flush(comp, arena, .main, main_progress_node);
                     ^
```

### CODE REVIEW

File: src/Compilation.zig
Comment (andrewrk) [inline src/Compilation.zig:3049]: If this function needs to allocate, then there's no point of it being separate from `getAllErrorsAlloc`. Just delete the function.

File: src/Zcu.zig
Comment (andrewrk) [inline src/Zcu.zig:175]: Does a dependency on `BoundedArray` really need to be dragged in? This could just be 3 flags, or even 0 flags with the (arguably trivial) logic repeated.

```zig
/// Rename to find all the logic that determines analysis roots.
pub const analysis_roots_logic = {};
```

elsewhere:

```zig
_ = analysis_roots_logic;
```

File: src/InternPool.zig
Comment (andrewrk) [inline src/InternPool.zig:250]: it's used on line 276. If it didn't take this argument, what would that line do instead?

File: src/InternPool.zig
Comment (andrewrk) [inline src/InternPool.zig:69]: Would you mind documenting the role of this type, as well as `TrackedInst`?

I'm struggling to follow along with most of this PR.

File: src/InternPool.zig
Comment (andrewrk) [inline src/InternPool.zig:276]: I'm looking at this arena field in the Local.mutate anonymous struct that is documented like this:

```zig
    /// This state is fully local to the owning thread and does not require any
    /// atomic access.
```

Could you document that arena field? I don't understand what is the lifetime of this arena, and why an arena is being used in the first place. Further, why is this the appropriate place for this map of... what?

File: src/InternPool.zig
Comment (andrewrk) [inline src/InternPool.zig:247]: what does it do, why is this needed, can you doc comment this please?

File: src/Zcu.zig
Comment (andrewrk) [inline src/Zcu.zig:3133]: suggestion per #363: 
* don't put the string "TODO" in the comment
* make a list of follow-up things to do in your PR writeup
* after merging, file issues for the follow-up things

File: src/Zcu.zig
Comment (andrewrk) [inline src/Zcu.zig:3110]: just to double check... this function is *only* called when a compile error occurs, right?

File: src/Zcu/PerThread.zig
Comment (andrewrk) [inline src/Zcu/PerThread.zig:39]: can you please add doc comments at the top of PerThread.zig? The role of this file is not intuitive to me

**Review by andrewrk — APPROVED** [review]
You can feel free to merge this when you feel it is ready. I have some requests mainly dealing with making the codebase more understandable. I think this logic has become quite complex, more than I am comfortable with, so let's start by at least documenting it more meticulously.

Let's make sure follow-up tasks are on the issue tracker rather than in TODO comments, please.

**Comment by mlugg** [comment]
Thanks for the review!

I'm currently going down a rabbit-hole of trying to make a more complex example work. Since there are no big conflicts coming in right now, I'll spend some time polishing this up -- implementing your review comments, and adding doc comments to explain things a little better. Sorry for the confusion -- this branch has had a lot of active experimentation, and my documentation efforts have fallen behind.

Something I might do is write one document which details the overall strategy for incremental compilation, since I have a suspicion I'm the only person with a solid understanding of it right now. If you're happy with that, would you prefer it go in the repo under `doc/`, or on the wiki?

File: src/Zcu.zig
Comment (mlugg) [reply src/Zcu.zig:3110]: Yep -- if there is an error in `failed_analysis`, we'll call `resolveReferences` from `Compilation`, which helps us (a) figure out the reference trace, and (b) determine if the failed `AnalUnit` is actually referenced in the latest update.

Except... mmm, I didn't actually put the call behind a condition. I'll do that.

File: src/Zcu.zig
Comment (mlugg) [reply src/Zcu.zig:175]: If you don't want to use `BoundedArray`, that's fair -- I can just replicate the logic. (FWIW, I don't really understand your general aversion to `BoundedArray` -- it's a pretty convenient abstraction at times IME)

**Comment by andrewrk** [comment]
> Something I might do is write one document which details the overall strategy for incremental compilation, since I have a suspicion I'm the only person with a solid understanding of it right now (my own fault!). If you're happy with that, would you prefer it go in the repo under `doc/`, or on the wiki?

That would be lovely. My suggestion is to put it somewhere other than the repo otherwise we'll get contributions to fix typos in it, and minor sentence rewordings that don't really matter.

**Comment by henrikolsen549** [comment]
@andrewrk, would love to see the docs in the repo:
- High-level design docs are extremely valuable. When placed in the repo, they are directly accessible.
- Historic versions of the docs are accessible in sync with the code via version control.
- Typos are highly irritating while reading, at least for detail-oriented personality types like me.
  Zig zen states: Favor reading code over writing code. I think this should also apply to docs.

**Comment by mlugg** [comment]
I've written up an informal high-level overview, mainly useful for existing compiler developers. @andrewrk, you may find this handy to read through at some point.

https://gist.github.com/mlugg/73b3e60c803006f3556d87c9ed3e8a0e

**Comment by mlugg** [comment]
The CI failure here is a timeout which is also happening on master, so I'm going to disregard it and merge now to avoid any conflicts with the Dwarf+link changes that I anticipate will continue coming in.

**Pull request merged**
